### PR TITLE
Timeout operator

### PR DESCRIPTION
**Pull request by tottakai**
Timeout operator as per #234 

@bontoJR said that we would be interested in doing the overload that takes the fallback observable (other) as a parameter.

### CODE REVIEW

**Comment by kzaher** [comment]
Hey @tottakai ,

will check it out over weekend, looks great on first read, but hard to really concentrate over week. 

Will check it out in more detail over weekend when I get some rest :)

**Comment by kzaher** [comment]
Hi @tottakai ,

I think the this code that you've provided for version without second observable is accurate :)

I think it would be great to reuse code for that `timeout` operator for both versions, with and without second observable (Swift generates huge binary dumps, and don't think it would make any significant performance difference :)

You can check out .NET code [here](https://github.com/Reactive-Extensions/Rx.NET/blob/master/Rx.NET/Source/System.Reactive.Linq/Reactive/Linq/Observable/Timeout.cs#L139) (it's probably close to what we need).

In that way, we could make `timeout` operator without second observable to be just a convenience for `timeout(delay, failWith(RxError.Timeout)`.

I think unit tests for it are [from here](https://github.com/Reactive-Extensions/Rx.NET/blob/859e6159cb07be67fd36b18c2ae2b9a62979cb6d/Rx.NET/Source/Tests.System.Reactive/Tests/Linq/ObservableTimeTest.cs#L6392) [to here](https://github.com/Reactive-Extensions/Rx.NET/blob/859e6159cb07be67fd36b18c2ae2b9a62979cb6d/Rx.NET/Source/Tests.System.Reactive/Tests/Linq/ObservableTimeTest.cs#L6702).

I think it would be wonderful to have fully compliant unit tests, don't think we have any reasons to diverge here :)

It's not too far from your original code, so I'm hoping you are ok with making those changes. Or if you don't have time, me or somebody else can do them ofc, no pressure :)

File: RxSwift/Observables/Implementations/Timeout.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    
+    func _synchronized_on(event: Event<E>) {
+        switch event {
+        case .Next:
+            forwardOn(event)
+            self._createTimeoutTimer()
+        case .Error:
+            forwardOn(event)
+            dispose()
+        case .Completed:
```
Comment (yury) [inline RxSwift/Observables/Implementations/Timeout.swift]: Can be written:
```
case .Error, .Completed:
  ...
```

**Comment by kzaher** [comment]
Ok, so this is a dupe of https://github.com/ReactiveX/RxSwift/pull/284 :)
I see @tottakai's commit there.

Closing this one then.

**Pull request closed without merging**